### PR TITLE
remove fp8 checkpoints for Attention

### PR DESCRIPTION
**Pull request by rachitgarg91**
# What does this PR do ?

Fix Checkpointing for FP8 Attention

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Jenkins CI
To run Jenkins, a NeMo User with write access must comment `jenkins` on the PR.

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Comment by ericharper** [comment]
@mikolajblaz can you review?

**Comment by ericharper** [comment]
jenkins

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py [hunk 1]
```diff
@@ -1722,6 +1724,14 @@
             if parallel_state.get_virtual_pipeline_model_parallel_world_size() is not None:
                 parallel_state.set_virtual_pipeline_model_parallel_rank(0)
 
+            def skip_fp8_load(x):
+                if isinstance(x, ShardedObject) and 'fused_attention' in x.key and '_extra_state' in x.key:
+                    x = LocalNonpersitentObject(x.data)  # use the FP8 state from initialization, not from ckpt
+                return x
+
+            if True:  # if self.cfg.skip_attn_fp8_load:
```
[the remark this answers is not in the document]
Comment (mikolajblaz) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:1732]: @rachitgarg91 can we configure it (like in the comment)?

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py [hunk 2]
```diff
@@ -1722,6 +1724,14 @@ def sharded_state_dict(self, prefix: str = '') -> Dict[str, Any]:
             if parallel_state.get_virtual_pipeline_model_parallel_world_size() is not None:
                 parallel_state.set_virtual_pipeline_model_parallel_rank(0)
 
+            def skip_fp8_load(x):
```
Comment (mikolajblaz) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:1727]: @rachitgarg91 Is this intended as a permanent fix?
If so, this should end up in the attention module or at least a TransformerLayer.
If it's temporary (I hope so?) please leave a comment that it's a WAR

Comment (mikolajblaz) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:1727]: also, does it make sense to move it to MCore?

**Comment by ericharper** [comment]
jenkins

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py [hunk 3]
```diff
@@ -1722,6 +1724,15 @@ def sharded_state_dict(self, prefix: str = '') -> Dict[str, Any]:
             if parallel_state.get_virtual_pipeline_model_parallel_world_size() is not None:
                 parallel_state.set_virtual_pipeline_model_parallel_rank(0)
 
+            # WAR: This is a temporary fix to skip loading FP8 parameters for Dot Product Attention
+            def skip_fp8_load(x):
+                if isinstance(x, ShardedObject) and 'fused_attention' in x.key and '_extra_state' in x.key:
+                    x = LocalNonpersitentObject(x.data)  # use the FP8 state from initialization, not from ckpt
+                return x
+
+            if self.cfg.fp8_dot_product_attention:
```
Comment (ericharper) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:1733]: This is going to break any config that doesn't have fp8_dot_product_attention attribute. Can you set a default?

**Review by mikolajblaz — APPROVED** [review]
thanks :+1:

**Comment by ShriyaPalsamudram** [comment]
jenkins

**Review by ShriyaPalsamudram — APPROVED** [review]
Approving based on Mikolaj's approval

**Pull request merged**